### ISSUE TITLE
On "staged" or atomic update operations

### ISSUE THREAD

**Issue by TravisCarden** [composer/composer#9779]
As mentioned in #9740, several CMSes are working to provide live, Composer-based automatic update functionality. In addition to being secure, we need updates to be essentially atomic so that a live application is never in an "in-between" state wherein requests are served with some `./vendor` files in the "before" state while others are in the "after" state. Obviously, no matter how efficient Composer is, no in-place operation that involves so many HTTP requests can ever achieve that in itself. Therefore we need some sort of "staged" update that, like a [blue-green deployment](https://martinfowler.com/bliki/BlueGreenDeployment.html), continues to serve the "before" state untouched while the update operation creates a separate "after" state and then simply swaps the two. As in #9740, **we're wondering if the Composer maintainers would be interested in either building such functionality into Composer or providing guidance on how it might be best accomplished in a plugin**. We have a few ideas for how the functionality might work in practice:

First of all, regardless of other architectural decisions, there must be essentially two `./vendor` directories during the update operation: the active "before" state and the inactive "after" state. Since the majority of users are likely to be on shared hosting, we can't assume that we'll have access to any directory outside the application root, including `/tmp`, so the "after" directory will need to be inside the root. Exactly where will depend on how we switch from "before" to "after":

The major possibilities I perceive boil down to 1) "swapping" the directories by renaming them, 2) copying differences from "after" to "before", or 3) using a symlink or similar to point `./vendor` from one to the other, or using the autoloader to change which one is "active".

changing which one is active using a pointer like a symlink or the autoloader.

1. We could rename the "before" directory from `./vendor` to something else and the "after" directory to `./vendor` to take its place.
   1. This approach has the benefit of complete atomicity. Unlike option two, there's never an instant in which the `./vendor` directory isn't *entirely* in the "before" state or *entirely* in the "after" state. Unlike option two, files don't have to be moved around after the update.
   2. On the other hand, there is an instant between the renamings, however brief, when there is no `./vendor` directory at all.
   3. This problem could be mitigated with a brief maintenance mode.
2. We could copy the differences from the "after" directory to the "before" directory with `rsync` or similar.
   1. This approach has the benefit that, unlike option one, there's always a `./vendor` directory present and it never moves.
   2. On the other hand, the operation is less atomic. `rsync` would be way faster than HTTP downloads, but if there are a lot of files to copy it may still take an unacceptably long time. What's more, `rsync` may not be available on some shared hosting, and alternative tools may be even less common.
   3. The non-atomicity problem could be mitigated with a maintenance mode. The inability to rely on the presence of `rsync` could be mitigated by using a Composer library that approximates it--though that could be slower yet.
3. In true blue-green fashion, we could could keep the two directories distinct and intact and either put a symlink at `./vendor` to point to whichever one is active or use the autoloader somehow to route conditionally between the two. While these approaches have some attractice benefits, they involve such drastic architectural changes that they probably aren't worth really considering. They're included merely for completeness.

Regardless of the overal architectural approach chosen, other questions must be answered:

1. Does this new "staging" behavior become the default? Or is it invoked via command line option (e.g., `--staged`) or an entirely separate command (e.g., `composer staged-update`)?
2. What events will be needed--for CMSes, for instance, to put their sites into and out of maintenance mode or to commit changes to Git?
3. Where does the "after" directory go until it is activated?
4. What happens to the "before" directory after the update is finished?
5. What about [Composer installers](https://getcomposer.org/doc/faqs/how-do-i-install-a-package-to-a-custom-path-for-my-framework.md) and installer paths?
6. What about preserving user uploads in the "before" directory, if necessary?
7. And of course, is this functionality to add to Composer itself? Or should it be offered as a contributed plugin?

**Comment by stof** [comment]
To me, this does not belong to composer.
A blue-green deployment is not only concerned about the `vendor` folder, but also about the application code. and once that enters into the equation, it becomes entirely out of the scope of composer, as it is not the one managing the deployment process (composer is run as part of deployment process, but it does not even need to be aware of a the rest).

>     6\. What about preserving user uploads in the "before" directory, if necessary?

Are you storing user uploads inside your `vendor` folder where Composer manages the content ? That looks wrong to me (and very fragile, as nothing in composer will guarantee you they are not removed)

**Comment by TravisCarden** [comment]
> Are you storing user uploads inside your vendor folder where Composer manages the content ?

No, I was brainstorming theoretical challenges... but you're right: it wouldn't be a good idea and Composer doesn't commit to supporting it, so there's no reason to be concerned with it. I removed it.

**Comment by effulgentsia** [comment]
> A blue-green deployment is not only concerned about the vendor folder, but also about the application code

That's correct. The OP describes the problem space very well, but I think is mistaken in referencing `./vendor`. This is about the entire project directory.

> once that enters into the equation, it becomes entirely out of the scope of composer, as it is not the one managing the deployment process (composer is run as part of deployment process, but it does not even need to be aware of a the rest)

I don't quite follow this reasoning, but maybe that's because I'm not familiar enough with what kinds of things are done within deployment processes that would be negatively affected by this.

I'll continue with this comment in case that's helpful even if it's already clear that this should be out of scope of composer.

> We could copy the differences from the "after" directory to the "before" directory with rsync or similar.

That's the option we've been exploring so far in https://github.com/php-tuf/staged-composer-update. The idea I currently have in mind for what it might look like (not fully implemented yet) is that that plugin (or Composer itself if it's interested in incorporating this feature) provides 3 commands: `staging-init`, `staging`, and `staging-deploy` (I welcome suggestions for better names or any other design improvement ideas).

- `composer staging-init` would copy everything other than certain "excluded" directories from the project directory to some staging directory (e.g., `.composer/staging` or similar).
- `composer staging ...` would run whatever the "..." is within the staging directory (similar to how `composer global ...` runs whatever the "..." is within the COMPOSER_HOME directory).
- `composer staging-deploy` would rsync (or similar) whatever is in the staging directory back into the project directory.

> What about preserving user uploads

This gets into the "excluded" directories mentioned above. The idea here is that there are certain directories within the project that get written to during the application's runtime. For example, the directory within a CMS where files uploaded by the CMS user get stored. These directories might therefore be written to at the same time that a `composer staging-init; ...; composer staging-deploy` process is occurring. So we want to exclude these directories from being overridden by a `composer staging-deploy`. We're thinking of managing this by marking these directories by placing something like an empty `.composer-staging-ignore` file in them and using that to build up the `--exclude` list to pass to rsync.

> we're wondering if the Composer maintainers would be interested in either building such functionality into Composer or providing guidance on how it might be best accomplished in a plugin

Yep, curious for your thoughts on whether any of this sounds interesting for Composer in general, or even if we keep it as a plugin, any feedback you might have about this problem space or approach. Thanks!

**Comment by Seldaek** [comment]
Overall this is pretty cool and I am glad to see interest in solving this "correctly" (I may be biased;) from CMSs, as we close in to Composer's 10th birthday :)

Going to try to answer the questions from OP first, at least those I have knowledge of:

> Does this new "staging" behavior become the default? Or is it invoked via command line option (e.g., --staged) or an entirely separate command (e.g., composer staged-update)?

Probably not. Really for dev time updates this is purely overhead/added complexity, and for the majority of projects which are not CMSs/apps doing updates in prod this makes no sense. As such I'd say it absolutely must be a separate process, and most likely remain outside/around Composer, as it seems like a huge problem space which we have little knowledge of and little resources available for. I think everyone will be happier if this remains in control of someone else, but you absolutely should seek to collaborate between CMSs to avoid rebuilding this if possible.

> Where does the "after" directory go until it is activated?
> What happens to the "before" directory after the update is finished?

What @effulgentsia described seems like a pretty good workflow to me. Definitely seems better considering installers and custom paths meaning that stuff can bleed out of the vendor dir, *especially* in CMSs which are the biggest users (and abusers.. I'll get back to that in a sec) of composer/installers.

> What about Composer installers and installer paths?

Once https://github.com/composer/composer/issues/9648 is resolved, I believe we'll have a good upgrade path in Composer 2.1 for projects who are interested in getting rid of their custom installers, in favor of using Composer\InstalledVersions to retrieve plugins of a given type, then fetch their install path.. I think the main reason for custom installers has been "how do I find which plugins have been dropped in dir X by the user", and while you  may need to keep that logic, the composer-installed ones should IMO remain in vendor and be loaded/detected via `InstalledVersions` instead of filesystem lookups. If you attain that goal you can then simplify the process above IMO by having a composer update purely constrained to the vendor dir and not reaching out all over the place. That'd mean you don't need to copy user/project files anymore, nor exclude them, etc.

> And of course, is this functionality to add to Composer itself? Or should it be offered as a contributed plugin?

As I wrote above, I think this should probably rather be a plugin or a separate tool entirely that calls Composer, which may be more practical than being a plugin, which comes with a few restrictions like the first install needs to happen before the plugin is available for example. And you're sharing memory space with Composer so you have to live with our dependency choices, that can also complicate things.

**Comment by TravisCarden** [comment]
Thanks for the helpful feedback, @Seldaek! (Happy early birthday. ;)

@effulgentsia and I discussed it, and we tend to agree that this makes more sense as a tool that *uses Composer* than as a feature *of Composer*. The approach we're currently investigating is a Symfony console command that requires `composer/composer`, sets the working directory (`--working-dir`) to our staging directory, and invokes it internally. (At a glance, `\Composer\Util\ProcessExecutor` looks like it might be a good place to start.) That would address most of the drawbacks you mention, and it would make it broadly usable as a project dependency or as a standalone tool.

Does that seem like sound reasoning?

**Comment by Seldaek** [comment]
Yes this seems reasonable to me except for:
- I'd probably use symfony/process natively rather than add a dependency on Composer's internals for process execution.
- Using Composer as a phar vs requiring it as a dependency leaves you a bit more flexible in terms of dependency management, as you won't include (and get stuck with) Composer's dependencies in your dependency graph.
- This approach requires a working php cli, which may be too big an assumption to make for some shared hosts? That's up to you of course, but I wonder if this will be a problem and in the end you may have to rather include Composer and run it in your web process. This isn't hard to do, even with the phar you could include stuff from the phar and then use something along the lines of https://github.com/composer/composer/blob/master/bin/composer#L62-L64 with a StringInput to run the command you want and use it as if it was run on CLI.

**Comment by TravisCarden** [comment]
Thanks, @Seldaek! That gives me enough to get started. What's your level of interest in the implementation? Would you like to give feedback on critical decisions, for instance? Would you just like to know when it's finished? etc.

**Comment by Seldaek** [comment]
I'm happy to review stuff if time allows, and to answer questions if you need guidance, in my experience this can often help avoid early mistakes which can have big repercussions down the line. So sure feel free to ping me.

**Comment by TravisCarden** [comment]
If it interests you, @Seldaek, https://github.com/php-tuf/composer-stager has a v0.1.0 release. 😃

**Comment by TravisCarden** [comment]
Today we released [php-tuf/composer-stager](https://github.com/php-tuf/composer-stager) v1.0.0-beta. As a beta, it's essentially feature complete and API-stable.

**Comment by TravisCarden** [comment]
Closing as essentially "solved". [php-tuf/composer-stager](https://github.com/php-tuf/composer-stager) has been stable for some time now.
